api/hasura/actions/_handlers/allocationCsv.ts

```diff
@@ -29,88 +32,120 @@ async function handler(req: VercelRequest, res: VercelResponse) {
-      operationName: 'allocationCsv_getGifts',
-    }
-  );
-  assert(circles_by_pk, 'No Circle Found');
-  const fixedPaymentsEnabled =
-    isFeatureEnabled('fixed_payments') &&
-    circles_by_pk.fixed_payment_token_type;
-  const users = circles_by_pk.users;
   const grant = payload.grant ?? epochObj.grant;
+
   const totalTokensSent = epochObj.token_gifts.length
     ? epochObj.token_gifts
         .map(g => g.tokens)
         .reduce((total, tokens) => tokens + total)
     : 0;
-  const dateRange = `${formatShortDateTime(
-    epochObj.start_date
-  )} - ${formatShortDateTime(epochObj.end_date)}`.replace(/,/g, '');
+  const circle = await getCircleDetails(
+    circle_id,
+    epochObj.id,
+    epochObj.end_date
+  );
+  assert(circle, 'No Circle Found');
+  const fixedPaymentsEnabled =
+    isFeatureEnabled('fixed_payments') && !!circle.fixed_payment_token_type;
+
+  const userValues = generateCsvValues(
+    circle,
+    form_gift_amount,
+    gift_token_symbol,
+    totalTokensSent,
+    fixedPaymentsEnabled,
+    circle.fixed_payment_token_type,
+    grant
+  );
+
   const headers = [
     'No',
     'name',
     'address',
     'received',
     'sent',
-    'epoch_number',
-    'Date',
+    'givers',
+    'percentage_of_give',
+    'circle_rewards',
+    'circle_rewards_token',
   ];
   if (fixedPaymentsEnabled) {
     headers.push('fixed_payment_amount');
     headers.push('fixed_payment_token_symbol');
   }
   if (grant) headers.push('Grant_amt');
-
   let csvText = `${headers.join(',')}\r\n`;
+  userValues.map(rowValues => {
+    csvText += `${rowValues.join(',')}\r\n`;
+  });
+  const fileName = `${epochObj.circle?.organization?.name}-${
+    epochObj.circle?.name
+  }-epoch-${epochObj.number}-date-${formatCustomDate(
+    epochObj.start_date,
+    'ddLLyy'
+  )}-${formatCustomDate(epochObj.end_date, 'ddLLyy')}.csv`;
+  const result = await uploadCsv(
+    `${circle_id}/${epochObj.id}/${uuidv4()}/${fileName}`,
+    csvText
+  );
+
+  res.status(200).json({
+    file: result.Location,
+  });
+}
+
+export function generateCsvValues(
+  circle: CircleDetails,
+  formGiftAmount: number,
+  giftTokenSymbol: string | undefined,
+  totalTokensSent: number,
+  fixedPaymentsEnabled: boolean,
+  fixedPaymentTokenType: string | undefined,
+  grant: number | undefined
+) {
+  assert(circle, 'No Circle Found');
+  assert(circle.epochs[0], 'No Epoch Found');
+
+  const distEpoch = circle.epochs[0];
+  const circleDist = distEpoch.distributions.find(
+    d =>
+      d.distribution_type === DISTRIBUTION_TYPE.GIFT ||
+      d.distribution_type === DISTRIBUTION_TYPE.COMBINED
+  );
+  const fixedDist = distEpoch.distributions.find(
+    d =>
+      d.distribution_type === DISTRIBUTION_TYPE.FIXED ||
+      d.distribution_type === DISTRIBUTION_TYPE.COMBINED
+  );
+
+  giftTokenSymbol = circleDist ? circleDist.vault.symbol : giftTokenSymbol;
+  const { users } = circle;
+
+  const userValues: (string | number)[][] = [];
   users.map((u, idx) => {
```

Comment:
this should probably be `.forEach` instead of `.map` since we're not using the returned result.